### ISSUE TITLE
Network policy rules do not apply to ClusterIP

### ISSUE THREAD

**Issue by thrawn997** [projectcalico/calico#1683]
## Expected Behavior

Applying a default deny rule for an application should block access to that application from the ClusterIP address. 

## Current Behavior

In a setup where ClusterIP is routeable on the network, Calico currently allows access to pods via any ClusterIP that selects them, even if there is no rule which allows that traffic.

## Your Environment

Kubernetes 1.9

**Comment by tmjd** [comment]
Could you add more details on your setup?  It would be good to know how you've deployed your K8s cluster, how you've deployed Calico, what version of Calico you're using, which backend datastore you're using, and what your policy and pod configuration are (or which demo/tutorial you're following).

A default deny rule will deny all incoming traffic to a Calico networked pod except traffic coming from the host it lives on.

**Comment by thrawn997** [comment]
tmjd,
We have Calico (quay.io/calico/cni:v1.11.2) deployed on bare-metal. 

You should be able to replicate this easily in any setup of calico on kubernetes.  Add a route from a host to a ClusterIP address of a service which selects a pod that is default denying ingress.   You will be able to reach the pod via this ClusterIP (perhaps clusterIP masquerades the source traffic as coming from the host?).

**Comment by tmjd** [comment]
So that CNI version is with Calico v2.6.4 to v2.6.7. Are you running calico-node in your setup also?  If using the etcd datastore you also need calico-kube-controllers. (Sorry if this is obvious but I want to make sure nothing obvious is missed.)

By ClusterIP, are your referring to the IP of a Kubernetes service?
Are you manually adding a route to a non-K8s host to reach a Cluster(Service)IP?  If you are manually adding a route to a host that is part of the K8s cluster that should not be needed and I think there is a setup issue.

I think you should verify you have the proper configuration as outlined [here](https://docs.projectcalico.org/v2.6/getting-started/kubernetes/installation/integration#configuring-kubernetes).
In addition to that you need kube-proxy configured with `--cluster-cidr=<Calico IP Pool>`?  That is needed for it to setup proper rules for forwarding Service IP traffic to the pods.
Does your service CIDR overlap with the cluster cidr because that will cause problems if they do?

**Comment by thrawn997** [comment]
We are using Kubernetes as the backend, and yes we use calico-node.  

By ClusterIP, yes, I mean the IP of a kubernetes service.  This is not a setup issue - this is intentional. Tim Hockin (from google) does the same for GCE customers: https://github.com/kubernetes/kubernetes/issues/27161 . There is nothing more inherently internal about a ClusterIP than a PodIP (both capture external traffic by default if that traffic has been routed to the host)

Our Service CIDR setup does not conflict with the  Cluster-cidr, which is correctly configured.

To Summarize:  Calico is working as intended.  The issue is that the intended design does not account for the fact that ClusterIPs are in some on-premise uses cases routable from outside of the cluster, and Calico currently does nothing to secure that traffic.

**Comment by tmjd** [comment]
Ah thank you for describing what you are doing a bit more.

That issue you linked to said that the traffic should be handled the same as a NodePort, have you tried the default deny when using a NodePort?  I'm asking because I have tested with a NodePort before and policy applies as expected when doing that.

Using a vagrant setup (with K8s v1.9.1) based off the one provided in the Calico docs, I just attempted what you suggested and saw traffic blocked as expected.  I added to my host machine a route (`sudo ip r add 10.100.0.0/24 via 172.18.18.102`) and then was able to access an exposed service (10.128.0.216), then I added the default deny in the [Simple Policy Demo](https://docs.projectcalico.org/v3.0/getting-started/kubernetes/tutorials/simple-policy#enable-isolation), was not able to access the service, then I added allow policy (using ipBlock) and was again able to access the service.

Maybe you could compare the configuration of the components in the Vagrant setup with your setup and locate some differences that are the cause of what you are seeing.  In particular you should check on your kube-proxy configuration since it is responsible for setting up the iptables rules for services.

**Comment by caseydavenport** [comment]
> perhaps clusterIP masquerades the source traffic as coming from the host?

I suspect this is probably what's happening here - Kubernetes NetworkPolicy mandates that a pod accept traffic from its host (in order to allow for kubelet liveness/readiness checks). The ClusterIP in this case will need to masquerade the traffic, and I don't think is smart enough to detect that the destination pod is local and thus no masq is required - the same behavior as NodePort without `externalTrafficPolicy: Local`.

However, if this _is_ the scenario we're seeing, I'd expect the traffic that goes `source`-> `clusterIP (nodeA)` -> `pod (nodeB)` to properly get denied. Have you checked that?

**Comment by thrawn997** [comment]
I'll run some tests next week and share the results. Thanks!

**Comment by caseydavenport** [comment]
@thrawn997 any luck with this guy?

**Comment by thrawn997** [comment]
Okay... for whatever reason, this is now working perfectly. I am not sure what the issue was, but since I can't reproduce I'll close.

Thanks for the help guys.

**Comment by samuela** [comment]
I'm also observing this behavior on GKE with Calico. I just deployed two services in different namespaces: foo and bar, along with a deny all policy:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default-deny
spec:
  podSelector: {}
  policyTypes:
  - Ingress
```
from https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors. But if I `kubectl exec -it foo-deployment-667cf67d6-kpm96 --namespace=foo -- /bin/sh` and run wget on the ClusterIP of the other service I'm still getting a response:
```
/ # wget -qO- 10.0.13.228
Hello, world!
Version: 2.0.0
Hostname: bar-deployment-67f7478c7-nqdb5
```
What's up with that? Am I misunderstanding the spec or is this a huge bug?

**Comment by samuela** [comment]
Update: `wget -qO- bar-service.bar` also works just as well.

**Comment by samuela** [comment]
Resolved: NetworkPolicys are namespaced. I was creating a NetworkPolicy in the default namespace and as a result it wasn't affecting pods in any other namespace. I guess it makes sense in hindsight, but identifying this took a significant amount of time.

**Comment by tungdam** [comment]
> > perhaps clusterIP masquerades the source traffic as coming from the host?
> 
> I suspect this is probably what's happening here - Kubernetes NetworkPolicy mandates that a pod accept traffic from its host (in order to allow for kubelet liveness/readiness checks). The ClusterIP in this case will need to masquerade the traffic, and I don't think is smart enough to detect that the destination pod is local and thus no masq is required - the same behavior as NodePort without `externalTrafficPolicy: Local`.
> 
> However, if this _is_ the scenario we're seeing, I'd expect the traffic that goes `source`-> `clusterIP (nodeA)` -> `pod (nodeB)` to properly get denied. Have you checked that?

I think I hit exactly this situation while applying network policy for our pod by ipBlock whitelist. 
The source IP of non-k8s node was changed to k8s-node. Using externalTrafficPolicy: Local helps with 1 replica only. 

Any suggestion to overcome this or we have to accept this, guys ? :(

**Comment by fasaxc** [comment]
Try setting the kube-proxy cluster CIDR.  That controls which IPs it does SNAT for.

**Comment by tungdam** [comment]
@fasaxc Thanks for your prompt response ! 
Our CIDR is set to clusterIP range already. 
Just to give you more info, our source IP was changed to k8s node IP. 
As we're using calico cluster IP advertising feature, the traffic from non-k8s-node to clusterIP will go like this:
non-k8s-node -> k8s-node ( as calico BGP next hop ) -> ( kube-proxy iptables rules ) -> pod.
Thus the src IP when it reaches pod will be k8s-node's IP , which doesn't match the network policy anymore.

**Comment by fasaxc** [comment]
@tungdam the (pod) cluster CIDR is not supposed to overlap with the service cluster IP CIDR

**Comment by tungdam** [comment]
I mean the cluster-cidr option in kube-proxy is set correctly ( to an ip range dedicated for service IP purpose ) 
Is that the cidr that you mentioned ?

**Comment by caseydavenport** [comment]
> I mean the cluster-cidr option in kube-proxy is set correctly ( to an ip range dedicated for service IP purpose )

The cluster-cidr option should be set to a dedicated range for pod IP addresses and should _not_ include the service IP range.

It's confusing for sure...

From [the docs](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/)

> The CIDR range of pods in the cluster. When configured, traffic sent to a Service cluster IP from outside this range will be masqueraded and traffic sent from pods to an external LoadBalancer IP will be directed to the respective cluster IP instead

**Comment by tungdam** [comment]
OMG, I misunderstood it completely. 
Thank you. Let me try to change this to see if it helps.

**Comment by kembox** [comment]
@fasaxc 
I corrected my ip ranges settings but still doesn't help. 
P/s: Sorry for keep this for too long, changing IP ranges if quite painful.

**Comment by tungdam** [comment]
@fasaxc I corrected my IP range settings but it still doesn't help
P/s: Sorry for keep this for too long, changing IP ranges is quite painful

**Comment by fasaxc** [comment]
@tungdam Please can you open a new issue with a full description of exactly the behaviour that you're seeing?  I.e. all the policies you're applying and the cluster/service CIDRs/IP pools you're using.

**Comment by tungdam** [comment]
@fasaxc : Ok, please continue [here](https://github.com/projectcalico/calico/issues/2660). 
Thanks
